File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/domain/oid4vp/DcqlMatcher.kt
package my.ssdid.sdk.domain.oid4vp

import kotlinx.serialization.json.*
import my.ssdid.sdk.domain.sdjwt.StoredSdJwtVc

class DcqlMatcher {

    fun match(dcql: JsonObject, credentials: List<StoredSdJwtVc>): List<MatchResult> {
        val credSpecs = dcql["credentials"]?.jsonArray ?: return emptyList()
        val results = mutableListOf<MatchResult>()

        for (spec in credSpecs) {
            val obj = spec.jsonObject
            val id = obj["id"]?.jsonPrimitive?.contentOrNull ?: continue
            val format = obj["format"]?.jsonPrimitive?.contentOrNull
            if (format != null && format != "vc+sd-jwt") continue

            val vctValues = obj["meta"]?.jsonObject
                ?.get("vct_values")?.jsonArray
                ?.map { it.jsonPrimitive.content }?.toSet()

            val claimsSpec = obj["claims"]?.jsonArray

            for (cred in credentials) {
                if (vctValues != null && cred.type !in vctValues) continue

                val (required, optional) = if (claimsSpec != null) {
                    extractClaims(claimsSpec, cred)
                } else {
                    cred.disclosableClaims to emptyList<String>()
                }

                results.add(MatchResult(cred, id, required, optional))
            }
        }
        return results
    }

    private fun extractClaims(
        claimsSpec: JsonArray,
        cred: StoredSdJwtVc
    ): Pair<List<String>, List<String>> {
        val required = mutableListOf<String>()
        val optional = mutableListOf<String>()
        for (claim in claimsSpec) {
            val obj = claim.jsonObject
            val paths = obj["path"]?.jsonArray?.map { it.jsonPrimitive.content } ?: continue
            val isOptional = obj["optional"]?.jsonPrimitive?.booleanOrNull == true
            val claimName = paths.firstOrNull() ?: continue
            if (claimName in cred.claims || claimName in cred.disclosableClaims) {
                if (isOptional) optional.add(claimName) else required.add(claimName)
            }
        }
        return required to optional
    }
}

File: sdk/ios/SsdidCore/Sources/SsdidCore/Domain/OpenId4Vp/DcqlMatcher.swift
import Foundation

/// Matches stored SD-JWT VCs against a DCQL (Digital Credentials Query Language) query.
final class DcqlMatcher {

    func match(
        dcql: [String: Any],
        credentials: [StoredSdJwtVc]
    ) -> [VpMatchResult] {
        guard let credSpecs = dcql["credentials"] as? [[String: Any]] else {
            return []
        }
        var results: [VpMatchResult] = []

        for spec in credSpecs {
            guard let id = spec["id"] as? String else { continue }

            if let format = spec["format"] as? String, format != "vc+sd-jwt" {
                continue
            }

            let vctValues: Set<String>? = {
                guard let meta = spec["meta"] as? [String: Any],
                      let values = meta["vct_values"] as? [String] else {
                    return nil
                }
                return Set(values)
            }()

            let claimsSpec = spec["claims"] as? [[String: Any]]

            for cred in credentials {
                if let vctValues = vctValues, !vctValues.contains(cred.type) {
                    continue
                }

                let (required, optional): ([String], [String])
                if let claimsSpec = claimsSpec {
                    (required, optional) = extractClaims(claimsSpec: claimsSpec, cred: cred)
                } else {
                    (required, optional) = (cred.disclosableClaims, [])
                }

                results.append(VpMatchResult(
                    credential: cred,
                    descriptorId: id,
                    requiredClaims: required,
                    optionalClaims: optional
                ))
            }
        }
        return results
    }

    // MARK: - Private

    private func extractClaims(
        claimsSpec: [[String: Any]],
        cred: StoredSdJwtVc
    ) -> (required: [String], optional: [String]) {
        var required: [String] = []
        var optional: [String] = []

        for claim in claimsSpec {
            guard let paths = claim["path"] as? [String],
                  let claimName = paths.first else {
                continue
            }
            let isOptional = claim["optional"] as? Bool == true

            if cred.claims[claimName] != nil || cred.disclosableClaims.contains(claimName) {
                if isOptional {
                    optional.append(claimName)
                } else {
                    required.append(claimName)
                }
            }
        }
        return (required, optional)
    }
}
